docs/proposals/koordlet/20230728-support-cold-memory-compute.md

```diff
@@ -0,0 +1,256 @@
+Define a function named GetIdleInfoFilePath(). It is uesd to return idleinfo file path.(node e.g. /sys/fs/cgroup/memory/memory.idle_stat)
+
+```go
+func GetIdleInfoFilePath(idleFileRelativePath string) string 
+```
+
+#### Report memory usage including hot page
+
+![image](../../images/support-cold-memory-2.svg)
+
+collectNodeMetric() is used to query node metirc and return CPU And MemUsed in pkg/koordlet/statesinformer/impl/states_nodemetirc.go file.
+
+We can report memory usage including hot pages in collectNodeMetric().
+
+The calculation formulas of node, pod and container are as follows.  
+
+node memory usage: nodeMemWithHotPageUasge=MemTotal-MemFree+NodeMeMWithHotPageSize
+
+pod memory usage: podMemWithHotPageUasge=m.InactiveAnon + m.ActiveAnon + m.Unevictable+PodMeMwithHotPageSize
+
+container memory usage: containerMemWithHotPageUasge=m.InactiveAnon + m.ActiveAnon + m.Unevictable+ContainerMeMwithHotPageSize
+
+The same process is executed pod informer to report memory usage. Do not repeat.
+
+#### Define cold memory API
+
+Provide memory collect policy access.
+
+Add field named MemoryCollectPolicy to represent which method is used to collect memory usage. Such as usageWithHotPageCache, usageWithoutPageCache,  usageWithPageCache.
+
+You can create a crd nodemeric resource and specify value of spec.CollectPolicy.MemoryCollectPolicy to start collect cold memory compute. 
+
+```go
+// NodeMetricCollectPolicy defines the Metric collection policy
+type NodeMetricCollectPolicy struct {
+	// AggregateDurationSeconds represents the aggregation period in seconds
+	AggregateDurationSeconds *int64 `json:"aggregateDurationSeconds,omitempty"`
+	// ReportIntervalSeconds represents the report period in seconds
+	ReportIntervalSeconds *int64 `json:"reportIntervalSeconds,omitempty"`
+	// NodeAggregatePolicy represents the target grain of node aggregated usage
+	NodeAggregatePolicy *AggregatePolicy `json:"nodeAggregatePolicy,omitempty"`
+	//MemoryWithHotPageCollectPolicy represents whether collet memory usage with hot page instead of memory usage without page cache
+	MemoryCollectPolicy string `json:"memoryCollectPolicy,omitempty"`
+}
+```
+## Alternatives
+
+**kidled**: kidled, an open source cold memory collection solution, is used by Anolis OS.
+
+https://github.com/alibaba/cloud-kernel/blob/linux-next/Documentation/vm/kidled.rst
+
+**kstaled**:  kstaled  is used by google linux kernel .It was designed by Michel. Michel's patch was developed on a early kernel version 3.0 and was similar to kidled. kstable also use /sys/kernel/mm/kstaled/scan_seconds and export idle_page_stats file. But kidled did not cherry pick the original kstaled's patch directly and made improvements. (e.g. design use_hierarchy )
+
+https://lwn.net/Articles/459269/
+
+https://lore.kernel.org/lkml/20110922161448.91a2e2b2.akpm@google.com/T/
+
+**DAMON**: DAMON was used  by amazon linux kernel.
+
+https://lwn.net/Articles/858682/
+
+why we choose kidled? The reasons are as below.
+
```

Comment:
Another question, on non-Anolis operating systems, what is the behavior of the program when this switch is turned on

Reply:
sepc.MemoryCollectPolicy in nodemetric to set cold page collector, sepc.MemoryCollectPolicy="" is off (os dose not support), sepc.MemoryCollectPolicy="kidled" (Anolis os suppoets),
sepc.MemoryCollectPolicy="kstaled" (Google os suppoets)
I am going to implement kidled first and provide interface to other cold collector

Reply:
@BUPT-wxq When the collect policy sets `coldpgs`(the user may not care about the detailed cold page technique), but the node does not start kidled, the koordlet should collect and report the memory usage without coldpg info instead of failing.